tiledb/api/src/column.rs

```diff
@@ -0,0 +1,509 @@
+            let len = if idx == num_offsets - 1 {
+                self.data().len() as u64 - *offset
+            } else {
+                offset - offsets[idx - 1]
+            };
+
+            let val = data[(*offset as usize)..((*offset + len) as usize)]
+                .to_vec()
+                .into_boxed_slice();
+
+            ret.push(val);
+        }
+
+        Ok(ret)
+    }
+
+    pub fn num_values(&self) -> u64 {
+        match self {
+            Column::Referenced(col) => col.num_values,
+            Column::Allocated(col) => col.num_values,
+        }
+    }
+
+    pub fn value_size(&self) -> u64 {
+        match self {
+            Column::Referenced(col) => col.value_size,
+            Column::Allocated(col) => col.value_size,
+        }
+    }
+
+    pub fn data(&self) -> &[u8] {
+        match self {
+            Column::Referenced(col) => col.data,
+            Column::Allocated(col) => &col.data[..],
+        }
+    }
+
+    pub fn offsets(&self) -> Option<&[u64]> {
+        match self {
+            Column::Referenced(col) => col.offsets,
+            Column::Allocated(col) => match &col.offsets {
+                Some(offsets) => unsafe {
+                    Some(std::slice::from_raw_parts(
+                        offsets.as_ptr(),
+                        self.num_values() as usize,
+                    ))
+                },
+                None => None,
+            },
+        }
+    }
+
+    pub fn is_allocated(&self) -> bool {
+        match self {
+            Column::Referenced(_) => false,
+            Column::Allocated(_) => true,
+        }
+    }
+
+    pub fn to_json(
+        &self,
+        datatype: Datatype,
+        cell_val_num: u32,
+    ) -> TileDBResult<JsonValue> {
+        let is_strings =
+            matches!(datatype, Datatype::StringAscii | Datatype::StringUtf8);
+
+        return if cell_val_num == u32::MAX && is_strings {
+            let result = self.to_string_vec(datatype, cell_val_num);
+            if result.is_ok() {
+                Ok(json!(result.unwrap()))
+            } else {
+                Ok(json!(self.to_boxed_u8_vec(datatype, cell_val_num)?))
+            }
+        } else if cell_val_num == u32::MAX {
+            Ok(json!(self.to_boxed_u8_vec(datatype, cell_val_num)?))
+        } else {
+            Ok(fn_typed!(self.to_vec, datatype, datatype, cell_val_num =>
+                json!(to_vec?)
+            ))
+        };
+    }
+}
+
+impl<'c1, 'c2> PartialEq<Column<'c2>> for Column<'c1> {
+    fn eq(&self, other: &Column<'c2>) -> bool {
+        if self.num_values() != other.num_values() {
+            return false;
+        }
+
+        if self.value_size() != other.value_size() {
+            return false;
+        }
+
+        if self.data() != other.data() {
+            return false;
+        }
+
+        let offsets_match = match (self.offsets(), other.offsets()) {
+            (Some(mine), Some(theirs)) => mine == theirs,
+            _ => false,
```

Comment:
Too strict for non-string columns.  `(None, None)` should also be a match.  You can probably just do `self.offset() == other.offsets()` to be honest

Reply:
Ooof, fair point on None, None being equal. I ran into clippy yelling at me to just directly compare slices and then validated that rustc emits memcmp in its assembly bits. However, when I tried figuring out if two Options are equal all of the suggestions were to use this pattern when its not certain if one or the other is None. Also, it had the standard "eventually `==` will be usable when random language feature X stabilizes" which was written three years ago and the feature is still not stable.